kafka/source.go

```diff
@@ -151,13 +158,17 @@ func (s *Source) Close() error {
 }
 
 func (s *Source) markState(msg *sarama.ConsumerMessage) {
+	s.stateLock.Lock()
+
 	partitions, ok := s.state[msg.Topic]
 	if !ok {
 		partitions = make(map[int32]int64)
 		s.state[msg.Topic] = partitions
 	}
 
 	partitions[msg.Partition] = msg.Offset
+
+	s.stateLock.Unlock()
```

Comment:
You could move it higher to unlock sooner

Reply:
This is all map access :)